File: reference/recipe.ts
import { type PatternId } from './blob47Pattern';
import { type NoiseId } from './patternNoise';
import { type RibbonId } from './patternRibbon';
import {
  DEFAULT_CELL_SCALE, MIN_CELL_SCALE, MAX_CELL_SCALE,
  DEFAULT_RIPPLE_SCALE, MIN_RIPPLE_SCALE, MAX_RIPPLE_SCALE,
  DEFAULT_GEO_SCALE, type TextureId
} from './patternTexture';

export interface Recipe {
  roleHex: {
    terrainA: string;
    terrainB: string;
    edge: string;
  };
  patternId: PatternId;
  edgeSeed: number;
  outlineWidth: number;
  bandSteps: number;
  hardEdgeB: boolean;
  transparentB: boolean;
  bandBias: number;
  customShadesHex: string[] | null;

  patternNoise: NoiseId[];
  patternNoiseSeed: number;
  patternNoiseStrength: number;

  ribbonAlgo: RibbonId;
  ribbonAmount: number;
  ribbonPeriod: number;
  ribbonShades: number;
  ribbonInvert: boolean;
  customRibbonHex: (string | undefined)[] | null;

  textureAlgoA: TextureId;
  textureAlgoB: TextureId;
  textureAmountA: number;
  textureAmountB: number;
  textureShadesA: number;
  textureShadesB: number;
  textureSeedA: number;
  textureSeedB: number;
  cellScaleA: number;
  cellScaleB: number;
  rippleScaleA: number;
  rippleScaleB: number;
  geoScaleA: number;
  geoScaleB: number;
  customTexHex: {
    terrainA: (string | undefined)[] | null;
    terrainB: (string | undefined)[] | null;
  };
}

export interface PresetItem {
  id: string;
  name: string;
  recipe: Recipe;
  savedAt?: string;
  isBuiltin?: boolean;
}

export const DEFAULT_RECIPE: Recipe = {
  roleHex: {
    terrainA: '#3a7fc9',
    terrainB: '#5da832',
    edge: '#e8d5a0',
  },
  patternId: 'rounded',
  edgeSeed: 0,
  outlineWidth: 2,
  bandSteps: 4,
  hardEdgeB: false,
  transparentB: false,
  bandBias: 0,
  customShadesHex: null,

  patternNoise: [],
  patternNoiseSeed: 1234,
  patternNoiseStrength: 0.15,

  ribbonAlgo: 'none',
  ribbonAmount: 0.25,
  ribbonPeriod: 4,
  ribbonShades: 2,
  ribbonInvert: false,
  customRibbonHex: null,

  textureAlgoA: 'none',
  textureAlgoB: 'none',
  textureAmountA: 0.35,
  textureAmountB: 0.35,
  textureShadesA: 2,
  textureShadesB: 2,
  textureSeedA: 1,
  textureSeedB: 1,
  cellScaleA: DEFAULT_CELL_SCALE,
  cellScaleB: DEFAULT_CELL_SCALE,
  rippleScaleA: DEFAULT_RIPPLE_SCALE,
  rippleScaleB: DEFAULT_RIPPLE_SCALE,
  geoScaleA: DEFAULT_GEO_SCALE,
  geoScaleB: DEFAULT_GEO_SCALE,
  customTexHex: {
    terrainA: null,
    terrainB: null,
  },
};

export const BUILTIN_PRESETS: PresetItem[] = [
  {
    id: 'builtin_waterfront',
    name: '水岸（默认）',
    isBuiltin: true,
    recipe: DEFAULT_RECIPE,
  },
];

const VALID_PATTERNS: Set<PatternId> = new Set<PatternId>([
  'square', 'sharp', 'rounded', 'wave', 'jagged', 'gravel',
  'boulder', 'thorn', 'coast', 'moss', 'billow',
]);

const VALID_NOISES: Set<NoiseId> = new Set<NoiseId>(['white', 'blue', 'ordered']);
const VALID_RIBBONS: Set<RibbonId> = new Set<RibbonId>([
  'none', 'bevel', 'dashes', 'ticks', 'beads', 'rope', 'wave', 'grain', 'speckle',
  'along_brick_wall', 'along_cobbles2', 'along_weave', 'along_stone_floor',
  'along_breeze_block', 'along_octagonal',
]);
const VALID_TEXTURES: Set<TextureId> = new Set<TextureId>([
  'none', 'white', 'blue', 'ordered', 'ripple', 'ripple_diag', 'cells',
  'breeze_block', 'brick_wall', 'cobbles2', 'brick_floor', 'hexagon',
  'isometric', 'isometric_grid', 'octagonal', 'square', 'weave',
  'paving', 'paving3', 'paving5', 'stone_floor', 'water', 'brick_bond',
  'field', 'rubble', 'nonslip',
]);

function isHexColor(str: unknown): str is string {
  return typeof str === 'string' && /^#[0-9a-fA-F]{6}$/.test(str);
}

function clamp(val: unknown, min: number, max: number, fallback: number): number {
  if (typeof val !== 'number' || isNaN(val)) return fallback;
  return Math.max(min, Math.min(max, val));
}

function clampInt(val: unknown, min: number, max: number, fallback: number): number {
  if (typeof val !== 'number' || isNaN(val)) return fallback;
  return Math.max(min, Math.min(max, Math.round(val)));
}

export function sanitizeRecipe(raw: unknown): Recipe {
  if (!raw || typeof raw !== 'object') return { ...DEFAULT_RECIPE };
  const obj = raw as Record<string, unknown>;

  const roleHexRaw = (obj.roleHex && typeof obj.roleHex === 'object') ? (obj.roleHex as Record<string, unknown>) : {};
  const terrainA = isHexColor(roleHexRaw.terrainA) ? roleHexRaw.terrainA : DEFAULT_RECIPE.roleHex.terrainA;
  const terrainB = isHexColor(roleHexRaw.terrainB) ? roleHexRaw.terrainB : DEFAULT_RECIPE.roleHex.terrainB;
  const edge = isHexColor(roleHexRaw.edge) ? roleHexRaw.edge : DEFAULT_RECIPE.roleHex.edge;

  const patternId = (typeof obj.patternId === 'string' && VALID_PATTERNS.has(obj.patternId as PatternId))
    ? (obj.patternId as PatternId)
    : DEFAULT_RECIPE.patternId;

  const edgeSeed = clampInt(obj.edgeSeed, 0, 99999, DEFAULT_RECIPE.edgeSeed);
  const outlineWidth = clampInt(obj.outlineWidth, 1, 4, DEFAULT_RECIPE.outlineWidth);
  const bandSteps = clampInt(obj.bandSteps, 3, 5, DEFAULT_RECIPE.bandSteps);
  const hardEdgeB = typeof obj.hardEdgeB === 'boolean' ? obj.hardEdgeB : DEFAULT_RECIPE.hardEdgeB;
  const transparentB = typeof obj.transparentB === 'boolean' ? obj.transparentB : DEFAULT_RECIPE.transparentB;
  const bandBias = clamp(obj.bandBias, -1, 1, DEFAULT_RECIPE.bandBias);

  let customShadesHex: string[] | null = null;
  if (Array.isArray(obj.customShadesHex) && obj.customShadesHex.length === bandSteps + 2) {
    customShadesHex = obj.customShadesHex.map((hex) => (isHexColor(hex) ? hex : ''));
    if (customShadesHex.some((h) => !h)) customShadesHex = null;
  }

  let patternNoise: NoiseId[] = DEFAULT_RECIPE.patternNoise;
  if (Array.isArray(obj.patternNoise)) {
    patternNoise = obj.patternNoise.filter((n): n is NoiseId => typeof n === 'string' && VALID_NOISES.has(n as NoiseId));
  }
  const patternNoiseSeed = clampInt(obj.patternNoiseSeed, 0, 99999, DEFAULT_RECIPE.patternNoiseSeed);
  const patternNoiseStrength = clamp(obj.patternNoiseStrength, 0, 2, DEFAULT_RECIPE.patternNoiseStrength);

  const ribbonAlgo = (typeof obj.ribbonAlgo === 'string' && VALID_RIBBONS.has(obj.ribbonAlgo as RibbonId))
    ? (obj.ribbonAlgo as RibbonId)
    : DEFAULT_RECIPE.ribbonAlgo;
  const ribbonAmount = clamp(obj.ribbonAmount, 0, 1, DEFAULT_RECIPE.ribbonAmount);
  const ribbonPeriod = clampInt(obj.ribbonPeriod, 1, 8, DEFAULT_RECIPE.ribbonPeriod);
  const ribbonShades = clampInt(obj.ribbonShades, 1, 4, DEFAULT_RECIPE.ribbonShades);
  const ribbonInvert = typeof obj.ribbonInvert === 'boolean' ? obj.ribbonInvert : DEFAULT_RECIPE.ribbonInvert;

  let customRibbonHex: (string | undefined)[] | null = null;
  if (Array.isArray(obj.customRibbonHex) && obj.customRibbonHex.length === ribbonShades + 1) {
    customRibbonHex = obj.customRibbonHex.map((h) => (isHexColor(h) ? h : undefined));
  }

  const textureAlgoA = (typeof obj.textureAlgoA === 'string' && VALID_TEXTURES.has(obj.textureAlgoA as TextureId))
    ? (obj.textureAlgoA as TextureId)
    : DEFAULT_RECIPE.textureAlgoA;
  const textureAlgoB = (typeof obj.textureAlgoB === 'string' && VALID_TEXTURES.has(obj.textureAlgoB as TextureId))
    ? (obj.textureAlgoB as TextureId)
    : DEFAULT_RECIPE.textureAlgoB;

  const textureAmountA = clamp(obj.textureAmountA, 0, 1, DEFAULT_RECIPE.textureAmountA);
  const textureAmountB = clamp(obj.textureAmountB, 0, 1, DEFAULT_RECIPE.textureAmountB);
  const textureShadesA = clampInt(obj.textureShadesA, 1, 4, DEFAULT_RECIPE.textureShadesA);
  const textureShadesB = clampInt(obj.textureShadesB, 1, 4, DEFAULT_RECIPE.textureShadesB);
  const textureSeedA = clampInt(obj.textureSeedA, 0, 99999, DEFAULT_RECIPE.textureSeedA);
  const textureSeedB = clampInt(obj.textureSeedB, 0, 99999, DEFAULT_RECIPE.textureSeedB);

  const cellScaleA = clampInt(obj.cellScaleA, MIN_CELL_SCALE, MAX_CELL_SCALE, DEFAULT_RECIPE.cellScaleA);
  const cellScaleB = clampInt(obj.cellScaleB, MIN_CELL_SCALE, MAX_CELL_SCALE, DEFAULT_RECIPE.cellScaleB);
  const rippleScaleA = clampInt(obj.rippleScaleA, MIN_RIPPLE_SCALE, MAX_RIPPLE_SCALE, DEFAULT_RECIPE.rippleScaleA);
  const rippleScaleB = clampInt(obj.rippleScaleB, MIN_RIPPLE_SCALE, MAX_RIPPLE_SCALE, DEFAULT_RECIPE.rippleScaleB);
  const geoScaleA = clampInt(obj.geoScaleA, 1, 8, DEFAULT_RECIPE.geoScaleA);
  const geoScaleB = clampInt(obj.geoScaleB, 1, 8, DEFAULT_RECIPE.geoScaleB);

  const customTexRaw = (obj.customTexHex && typeof obj.customTexHex === 'object')
    ? (obj.customTexHex as Record<string, unknown>)
    : {};
  let customTexA: (string | undefined)[] | null = null;
  if (Array.isArray(customTexRaw.terrainA) && customTexRaw.terrainA.length === textureShadesA + 1) {
    customTexA = customTexRaw.terrainA.map((h) => (isHexColor(h) ? h : undefined));
  }
  let customTexB: (string | undefined)[] | null = null;
  if (Array.isArray(customTexRaw.terrainB) && customTexRaw.terrainB.length === textureShadesB + 1) {
    customTexB = customTexRaw.terrainB.map((h) => (isHexColor(h) ? h : undefined));
  }

  return {
    roleHex: { terrainA, terrainB, edge },
    patternId,
    edgeSeed,
    outlineWidth,
    bandSteps,
    hardEdgeB,
    transparentB,
    bandBias,
    customShadesHex,

    patternNoise,
    patternNoiseSeed,
    patternNoiseStrength,

    ribbonAlgo,
    ribbonAmount,
    ribbonPeriod,
    ribbonShades,
    ribbonInvert,
    customRibbonHex,

    textureAlgoA,
    textureAlgoB,
    textureAmountA,
    textureAmountB,
    textureShadesA,
    textureShadesB,
    textureSeedA,
    textureSeedB,
    cellScaleA,
    cellScaleB,
    rippleScaleA,
    rippleScaleB,
    geoScaleA,
    geoScaleB,
    customTexHex: {
      terrainA: customTexA,
      terrainB: customTexB,
    },
  };
}

File: src/model/recipe.cpp
#include "recipe.h"
#include "pattern/js_math.h"
#include "pattern/pattern_paint.h"
#include <unordered_set>
#include <regex>
#include <cmath>

namespace atm {

static const Recipe DEFAULT_RECIPE_INSTANCE;

namespace {

/** Grow or shrink a sparse override array; new slots mean "not overridden". */
void resize_sparse(std::optional<std::vector<std::optional<std::string>>>& arr, size_t want) {
    if (!arr.has_value()) return;
    arr->resize(want);
}

} // namespace

void sync_band_overrides(Recipe& r) {
    if (!r.customShadesHex.has_value()) return;

    const size_t want = static_cast<size_t>(r.bandSteps) + 2;
    auto& arr = *r.customShadesHex;
    if (arr.size() == want) return;

    if (arr.size() > want) {
        arr.resize(want);
        return;
    }

    // customShadesHex is all-or-nothing: an empty entry invalidates the whole
    // array, so a grown level cannot be left blank. Fill it with the shade the
    // level would have had with no override at all.
    RoleColours colours{
        parse_hex_colour(r.roleHex.terrainA),
        parse_hex_colour(r.roleHex.terrainB),
        parse_hex_colour(r.roleHex.edge)
    };
    const std::vector<RGB> computed = pattern_ramp(colours, r.bandSteps);

    const size_t first_new = arr.size();
    arr.resize(want);
    for (size_t i = first_new; i < want; ++i) {
        arr[i] = (i < computed.size()) ? to_hex_colour(computed[i])
                                       : to_hex_colour(colours.terrainA);
    }
}

void sync_ribbon_overrides(Recipe& r) {
    resize_sparse(r.customRibbonHex, static_cast<size_t>(r.ribbonShades) + 1);
}

void sync_texture_overrides(Recipe& r) {
    resize_sparse(r.customTexHexA, static_cast<size_t>(r.textureShadesA) + 1);
    resize_sparse(r.customTexHexB, static_cast<size_t>(r.textureShadesB) + 1);
}

void sync_all_overrides(Recipe& r) {
    sync_band_overrides(r);
    sync_ribbon_overrides(r);
    sync_texture_overrides(r);
}

const Recipe& get_default_recipe() {
    return DEFAULT_RECIPE_INSTANCE;
}

static bool is_hex_color(const std::string& str) {
    if (str.length() != 7 || str[0] != '#') return false;
    for (int i = 1; i < 7; ++i) {
        char c = str[i];
        if (!((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F'))) {
            return false;
        }
    }
    return true;
}

static int clamp_int(int val, int min_val, int max_val, int fallback) {
    if (std::isnan(static_cast<float>(val))) return fallback;
    return std::max(min_val, std::min(max_val, val));
}

static double clamp_double(double val, double min_val, double max_val, double fallback) {
    if (!std::isfinite(val)) return fallback;
    return std::max(min_val, std::min(max_val, val));
}

Recipe sanitize_recipe(const nlohmann::json& raw) {
    if (!raw.is_object()) return DEFAULT_RECIPE_INSTANCE;

    Recipe r;

    if (raw.contains("roleHex") && raw["roleHex"].is_object()) {
        const auto& rh = raw["roleHex"];
        if (rh.contains("terrainA") && rh["terrainA"].is_string() && is_hex_color(rh["terrainA"].get<std::string>())) {
            r.roleHex.terrainA = rh["terrainA"].get<std::string>();
        }
        if (rh.contains("terrainB") && rh["terrainB"].is_string() && is_hex_color(rh["terrainB"].get<std::string>())) {
            r.roleHex.terrainB = rh["terrainB"].get<std::string>();
        }
        if (rh.contains("edge") && rh["edge"].is_string() && is_hex_color(rh["edge"].get<std::string>())) {
            r.roleHex.edge = rh["edge"].get<std::string>();
        }
    }

    static const std::unordered_set<std::string> VALID_PATTERNS = {
        "square", "sharp", "rounded", "wave", "jagged", "gravel",
        "boulder", "thorn", "coast", "moss", "billow"
    };

    if (raw.contains("patternId") && raw["patternId"].is_string()) {
        std::string pid = raw["patternId"].get<std::string>();
        if (VALID_PATTERNS.count(pid)) {
            r.patternId = pid;
        }
    }

    if (raw.contains("edgeSeed") && raw["edgeSeed"].is_number()) {
        r.edgeSeed = clamp_int(raw["edgeSeed"].get<int>(), 0, 99999, DEFAULT_RECIPE_INSTANCE.edgeSeed);
    }
    if (raw.contains("outlineWidth") && raw["outlineWidth"].is_number()) {
        r.outlineWidth = clamp_int(raw["outlineWidth"].get<int>(), 1, 4, DEFAULT_RECIPE_INSTANCE.outlineWidth);
    }
    if (raw.contains("bandSteps") && raw["bandSteps"].is_number()) {
        r.bandSteps = clamp_int(raw["bandSteps"].get<int>(), 3, 5, DEFAULT_RECIPE_INSTANCE.bandSteps);
    }
    if (raw.contains("hardEdgeB") && raw["hardEdgeB"].is_boolean()) {
        r.hardEdgeB = raw["hardEdgeB"].get<bool>();
    }
    if (raw.contains("transparentB") && raw["transparentB"].is_boolean()) {
        r.transparentB = raw["transparentB"].get<bool>();
    }
    if (raw.contains("bandBias") && raw["bandBias"].is_number()) {
        r.bandBias = clamp_double(raw["bandBias"].get<double>(), -1.0, 1.0, DEFAULT_RECIPE_INSTANCE.bandBias);
    }

    if (raw.contains("customShadesHex") && raw["customShadesHex"].is_array()) {
        const auto& arr = raw["customShadesHex"];
        if (static_cast<int>(arr.size()) == r.bandSteps + 2) {
            std::vector<std::string> custom;
            bool valid = true;
            for (const auto& item : arr) {
                if (item.is_string() && is_hex_color(item.get<std::string>())) {
                    custom.push_back(item.get<std::string>());
                } else {
                    valid = false;
                    break;
                }
            }
            if (valid) r.customShadesHex = custom;
        }
    }

    if (raw.contains("patternNoise") && raw["patternNoise"].is_array()) {
        r.patternNoise.clear();
        for (const auto& item : raw["patternNoise"]) {
            if (item.is_string()) {
                std::string s = item.get<std::string>();
                if (s == "white" || s == "blue" || s == "ordered") {
                    r.patternNoise.push_back(parse_noise_id(s));
                }
            }
        }
    }
    if (raw.contains("patternNoiseSeed") && raw["patternNoiseSeed"].is_number()) {
        r.patternNoiseSeed = clamp_int(raw["patternNoiseSeed"].get<int>(), 0, 99999, DEFAULT_RECIPE_INSTANCE.patternNoiseSeed);
    }
    if (raw.contains("patternNoiseStrength") && raw["patternNoiseStrength"].is_number()) {
        r.patternNoiseStrength = clamp_double(raw["patternNoiseStrength"].get<double>(), 0.0, 2.0, DEFAULT_RECIPE_INSTANCE.patternNoiseStrength);
    }

    static const std::unordered_set<std::string> VALID_RIBBONS = {
        "none", "bevel", "dashes", "ticks", "beads", "rope", "wave", "grain", "speckle",
        "along_brick_wall", "along_cobbles2", "along_weave", "along_stone_floor",
        "along_breeze_block", "along_octagonal"
    };

    if (raw.contains("ribbonAlgo") && raw["ribbonAlgo"].is_string()) {
        std::string ral = raw["ribbonAlgo"].get<std::string>();
        if (VALID_RIBBONS.count(ral)) r.ribbonAlgo = ral;
    }
    if (raw.contains("ribbonAmount") && raw["ribbonAmount"].is_number()) {
        r.ribbonAmount = clamp_double(raw["ribbonAmount"].get<double>(), 0.0, 1.0, DEFAULT_RECIPE_INSTANCE.ribbonAmount);
    }
    if (raw.contains("ribbonPeriod") && raw["ribbonPeriod"].is_number()) {
        r.ribbonPeriod = clamp_int(raw["ribbonPeriod"].get<int>(), 1, 8, DEFAULT_RECIPE_INSTANCE.ribbonPeriod);
    }
    if (raw.contains("ribbonShades") && raw["ribbonShades"].is_number()) {
        r.ribbonShades = clamp_int(raw["ribbonShades"].get<int>(), 1, 4, DEFAULT_RECIPE_INSTANCE.ribbonShades);
    }
    if (raw.contains("ribbonInvert") && raw["ribbonInvert"].is_boolean()) {
        r.ribbonInvert = raw["ribbonInvert"].get<bool>();
    }

    if (raw.contains("customRibbonHex") && raw["customRibbonHex"].is_array()) {
        const auto& arr = raw["customRibbonHex"];
        if (static_cast<int>(arr.size()) == r.ribbonShades + 1) {
            std::vector<std::optional<std::string>> custom;
            for (const auto& item : arr) {
                if (item.is_string() && is_hex_color(item.get<std::string>())) {
                    custom.push_back(item.get<std::string>());
                } else {
                    custom.push_back(std::nullopt);
                }
            }
            r.customRibbonHex = custom;
        }
    }

    static const std::unordered_set<std::string> VALID_TEXTURES = {
        "none", "white", "blue", "ordered", "ripple", "ripple_diag", "cells",
        "breeze_block", "brick_wall", "cobbles2", "brick_floor", "hexagon",
        "isometric", "isometric_grid", "octagonal", "square", "weave",
        "paving", "paving3", "paving5", "stone_floor", "water", "brick_bond",
        "field", "rubble", "nonslip"
    };

    if (raw.contains("textureAlgoA") && raw["textureAlgoA"].is_string()) {
        std::string tal = raw["textureAlgoA"].get<std::string>();
        if (VALID_TEXTURES.count(tal)) r.textureAlgoA = tal;
    }
    if (raw.contains("textureAlgoB") && raw["textureAlgoB"].is_string()) {
        std::string tal = raw["textureAlgoB"].get<std::string>();
        if (VALID_TEXTURES.count(tal)) r.textureAlgoB = tal;
    }

    if (raw.contains("textureAmountA") && raw["textureAmountA"].is_number()) {
        r.textureAmountA = clamp_double(raw["textureAmountA"].get<double>(), 0.0, 1.0, DEFAULT_RECIPE_INSTANCE.textureAmountA);
    }
    if (raw.contains("textureAmountB") && raw["textureAmountB"].is_number()) {
        r.textureAmountB = clamp_double(raw["textureAmountB"].get<double>(), 0.0, 1.0, DEFAULT_RECIPE_INSTANCE.textureAmountB);
    }
    if (raw.contains("textureShadesA") && raw["textureShadesA"].is_number()) {
        r.textureShadesA = clamp_int(raw["textureShadesA"].get<int>(), 1, 4, DEFAULT_RECIPE_INSTANCE.textureShadesA);
    }
    if (raw.contains("textureShadesB") && raw["textureShadesB"].is_number()) {
        r.textureShadesB = clamp_int(raw["textureShadesB"].get<int>(), 1, 4, DEFAULT_RECIPE_INSTANCE.textureShadesB);
    }
    if (raw.contains("textureSeedA") && raw["textureSeedA"].is_number()) {
        r.textureSeedA = clamp_int(raw["textureSeedA"].get<int>(), 0, 99999, DEFAULT_RECIPE_INSTANCE.textureSeedA);
    }
    if (raw.contains("textureSeedB") && raw["textureSeedB"].is_number()) {
        r.textureSeedB = clamp_int(raw["textureSeedB"].get<int>(), 0, 99999, DEFAULT_RECIPE_INSTANCE.textureSeedB);
    }

    if (raw.contains("cellScaleA") && raw["cellScaleA"].is_number()) {
        r.cellScaleA = clamp_int(raw["cellScaleA"].get<int>(), 2, 8, DEFAULT_RECIPE_INSTANCE.cellScaleA);
    }
    if (raw.contains("cellScaleB") && raw["cellScaleB"].is_number()) {
        r.cellScaleB = clamp_int(raw["cellScaleB"].get<int>(), 2, 8, DEFAULT_RECIPE_INSTANCE.cellScaleB);
    }
    if (raw.contains("rippleScaleA") && raw["rippleScaleA"].is_number()) {
        r.rippleScaleA = clamp_int(raw["rippleScaleA"].get<int>(), 2, 8, DEFAULT_RECIPE_INSTANCE.rippleScaleA);
    }
    if (raw.contains("rippleScaleB") && raw["rippleScaleB"].is_number()) {
        r.rippleScaleB = clamp_int(raw["rippleScaleB"].get<int>(), 2, 8, DEFAULT_RECIPE_INSTANCE.rippleScaleB);
    }
    if (raw.contains("geoScaleA") && raw["geoScaleA"].is_number()) {
        r.geoScaleA = clamp_int(raw["geoScaleA"].get<int>(), 1, 8, DEFAULT_RECIPE_INSTANCE.geoScaleA);
    }
    if (raw.contains("geoScaleB") && raw["geoScaleB"].is_number()) {
        r.geoScaleB = clamp_int(raw["geoScaleB"].get<int>(), 1, 8, DEFAULT_RECIPE_INSTANCE.geoScaleB);
    }

    if (raw.contains("customTexHex") && raw["customTexHex"].is_object()) {
        const auto& ctex = raw["customTexHex"];
        if (ctex.contains("terrainA") && ctex["terrainA"].is_array()) {
            const auto& arr = ctex["terrainA"];
            if (static_cast<int>(arr.size()) == r.textureShadesA + 1) {
                std::vector<std::optional<std::string>> custom;
                for (const auto& item : arr) {
                    if (item.is_string() && is_hex_color(item.get<std::string>())) {
                        custom.push_back(item.get<std::string>());
                    } else {
                        custom.push_back(std::nullopt);
                    }
                }
                r.customTexHexA = custom;
            }
        }
        if (ctex.contains("terrainB") && ctex["terrainB"].is_array()) {
            const auto& arr = ctex["terrainB"];
            if (static_cast<int>(arr.size()) == r.textureShadesB + 1) {
                std::vector<std::optional<std::string>> custom;
                for (const auto& item : arr) {
                    if (item.is_string() && is_hex_color(item.get<std::string>())) {
                        custom.push_back(item.get<std::string>());
                    } else {
                        custom.push_back(std::nullopt);
                    }
                }
                r.customTexHexB = custom;
            }
        }
    }

    return r;
}

Recipe recipe_from_json(const nlohmann::json& j) {
    return sanitize_recipe(j);
}

nlohmann::json recipe_to_json(const Recipe& r) {
    nlohmann::json j;
    j["roleHex"]["terrainA"] = r.roleHex.terrainA;
    j["roleHex"]["terrainB"] = r.roleHex.terrainB;
    j["roleHex"]["edge"] = r.roleHex.edge;

    j["patternId"] = r.patternId;
    j["edgeSeed"] = r.edgeSeed;
    j["outlineWidth"] = r.outlineWidth;
    j["bandSteps"] = r.bandSteps;
    j["hardEdgeB"] = r.hardEdgeB;
    j["transparentB"] = r.transparentB;
    j["bandBias"] = r.bandBias;

    if (r.customShadesHex.has_value()) {
        j["customShadesHex"] = *r.customShadesHex;
    } else {
        j["customShadesHex"] = nullptr;
    }

    nlohmann::json noiseArr = nlohmann::json::array();
    for (NoiseId nid : r.patternNoise) {
        noiseArr.push_back(noise_id_to_string(nid));
    }
    j["patternNoise"] = noiseArr;
    j["patternNoiseSeed"] = r.patternNoiseSeed;
    j["patternNoiseStrength"] = r.patternNoiseStrength;

    j["ribbonAlgo"] = r.ribbonAlgo;
    j["ribbonAmount"] = r.ribbonAmount;
    j["ribbonPeriod"] = r.ribbonPeriod;
    j["ribbonShades"] = r.ribbonShades;
    j["ribbonInvert"] = r.ribbonInvert;

    if (r.customRibbonHex.has_value()) {
        nlohmann::json arr = nlohmann::json::array();
        for (const auto& item : *r.customRibbonHex) {
            if (item.has_value()) arr.push_back(*item);
            else arr.push_back(nullptr);
        }
        j["customRibbonHex"] = arr;
    } else {
        j["customRibbonHex"] = nullptr;
    }

    j["textureAlgoA"] = r.textureAlgoA;
    j["textureAlgoB"] = r.textureAlgoB;
    j["textureAmountA"] = r.textureAmountA;
    j["textureAmountB"] = r.textureAmountB;
    j["textureShadesA"] = r.textureShadesA;
    j["textureShadesB"] = r.textureShadesB;
    j["textureSeedA"] = r.textureSeedA;
    j["textureSeedB"] = r.textureSeedB;
    j["cellScaleA"] = r.cellScaleA;
    j["cellScaleB"] = r.cellScaleB;
    j["rippleScaleA"] = r.rippleScaleA;
    j["rippleScaleB"] = r.rippleScaleB;
    j["geoScaleA"] = r.geoScaleA;
    j["geoScaleB"] = r.geoScaleB;

    nlohmann::json ctex;
    if (r.customTexHexA.has_value()) {
        nlohmann::json arr = nlohmann::json::array();
        for (const auto& item : *r.customTexHexA) {
            if (item.has_value()) arr.push_back(*item);
            else arr.push_back(nullptr);
        }
        ctex["terrainA"] = arr;
    } else {
        ctex["terrainA"] = nullptr;
    }

    if (r.customTexHexB.has_value()) {
        nlohmann::json arr = nlohmann::json::array();
        for (const auto& item : *r.customTexHexB) {
            if (item.has_value()) arr.push_back(*item);
            else arr.push_back(nullptr);
        }
        ctex["terrainB"] = arr;
    } else {
        ctex["terrainB"] = nullptr;
    }
    j["customTexHex"] = ctex;

    return j;
}

} // namespace atm
